substrate/frame/assets-freezer/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Assets Freezer Pallet
+//!
+//! A pallet capable of freezing fungibles from `pallet-assets`.
+//!
+//! > Made with *Substrate*, for *Polkadot*.
+//!
+//! [![github]](https://github.com/paritytech/polkadot-sdk/tree/master/substrate/frame/examples/basic)
+//! [![polkadot]](https://polkadot.network)
+//!
+//! [polkadot]: https://img.shields.io/badge/polkadot-E6007A?style=for-the-badge&logo=polkadot&logoColor=white
+//! [github]: https://img.shields.io/badge/github-8da0cb?style=for-the-badge&labelColor=555555&logo=github
+//!
+//! ## Pallet API
+//!
+//! See the [`pallet`] module for more information about the interfaces this pallet exposes,
+//! including its configuration trait, dispatchables, storage items, events and errors.
+//!
+//! ## Overview
+//!
+//! This pallet provides the following functionality:
+//!
+//! - Pallet hooks that implement custom logic to let `pallet-assets` know whether an balance is
+//!   frozen for an account on a given asset (see: [`pallet_assets::FrozenBalance`]).
+//! - An implementation of the fungibles
+//!   [inspect][`frame_support::traits::fungibles::InspectFreeze`] and the
+//!   [mutation][`frame_support::traits::fungibles::InspectFreeze`] APIs for freezes.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::pallet_prelude::DispatchResult;
+use frame_system::pallet_prelude::BlockNumberFor;
+use sp_runtime::{
+	traits::{Saturating, Zero},
+	BoundedSlice,
+};
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+mod impls;
+mod types;
+pub use types::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	use codec::FullCodec;
+	use core::fmt::Debug;
+	use frame_support::{pallet_prelude::*, traits::VariantCount, BoundedVec};
+
+	#[pallet::config]
+	pub trait Config<I: 'static = ()>: frame_system::Config + pallet_assets::Config<I> {
+		/// The overarching freeze reason.
+		type RuntimeFreezeReason: VariantCount
+			+ FullCodec
+			+ TypeInfo
+			+ PartialEq
+			+ Ord
+			+ MaxEncodedLen
+			+ Clone
+			+ Debug
+			+ 'static;
```

Comment:
Does this work?

```suggestion
        /// The overarching freeze reason.
		#[pallet::no_default_bounds]
		type RuntimeFreezeReason: VariantCount;
```

Reply:
Now that I think, this is related as to why `pallet-balances` uses `FreezeIdentifier` instead of `RuntimeFreezeReason` to store actual freezes. IMHO, a better solution for this storage long-term is to follow the same logic that pallet uses.